Add GPU support for DNNL endpoint

georgen117 opened the pull request:
**Description**: Describe your changes.

This enables the DNNL endpoint to use Intel GPU when training a model with Relu, Convolution, or MaxPool operators.

**Motivation and Context**
- Why is this change required? What problem does it solve?

A recent commit introduced ReluGrad, ConvGrad, and MaxPoolGrad ops for ORT training when using the DNNL endpoint. This makes it possible to use those ops with an Intel GPU.

jywu-msft commented:
@mrry can you help take a look since you reviewed the previous PR's?

mrry commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:
```diff
@@ -314,33 +312,50 @@ class DnnlConvGrad : public DnnlKernel {
 
     diff_dst_mem_ = onnxruntime::make_unique<dnnl::memory>(
-        dnnl::memory(conv_bwd_weights_pd_.get()->diff_dst_desc(), engine_to_use, nullptr));
+        dnnl::memory(conv_bwd_weights_pd_.get()->diff_dst_desc(), cpu_engine, nullptr));
```
Checking my understanding here: is it intentional that these will be created for the CPU engine even when a GPU is available?

(I would have expected e.g. `diff_dst_mem_` to contain a reference to GPU memory when a GPU is available, and the same member to point to CPU memory otherwise. This seems to be how `primitive_dst_mem_` is handled. But I'm not familiar with how this is actually implemented....)

georgen117 replied:
This is describing the memory block that will be bound to in the Bind function. This is memory is from OnnxRuntime and exists in the CPU ram, not the GPU ram.  This memory will be copied to the GPU using the reorder below (Lines 327-328). So the cpu_engine is done purposely and is required.

mrry replied:
Thanks for clarifying!

mrry commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:
```diff
@@ -6,6 +6,33 @@
 
+
+/*
+ConvGrad: (According to OnnxRuntime discovered using code inspection and Onnx documentation)
```
Should this be "MaxPoolGrad"?

georgen117 replied:
fixed

mrry commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:
```diff
@@ -314,33 +312,50 @@ class DnnlConvGrad : public DnnlKernel {
 
     src_mem_ = onnxruntime::make_unique<dnnl::memory>(
-        dnnl::memory(conv_bwd_weights_pd_.get()->src_desc(), engine_to_use, nullptr));
+        dnnl::memory(conv_bwd_weights_pd_.get()->src_desc(), cpu_engine, nullptr));
```
Style nit: for all of these `make_unique` calls, I think you shouldn't need to write the type name type. e.g. The following should work:

```c++
src_mem_ = onnxruntime::make_unique<dnnl::memory>(conv_bwd_weights_pd_.get()->src_desc(), cpu_engine, nullptr);
```

georgen117 replied:
@mrry This is one of those things that stuck around because I was following the style of past developers.  I looked at the code in other files, for this endpoint and this style nit is everywhere. I am already working on a cleanup commit that only does code cleanup.  This would be better addressed in that commit instead of this commit.

mrry replied:
Sounds great, thanks.

mrry commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:
```diff
@@ -144,82 +148,115 @@ class DnnlMaxPoolGrad : public DnnlKernel {
+      // reorder source memory for best performance (AVX512);
+      if (primitive_dst_desc_ != source_desc_) {
+        //dnnl::memory::dims src_dims(xgrad_shape_.GetDims().begin(), xgrad_shape_.GetDims().end());
+        auto pd = dnnl::memory::desc(source_desc_);
```
Style nit: The following would be a little more compact... (I see this pattern in a few places).

```c++
dnnl::memory::desc pd(source_desc_);
```

georgen117 replied:
Thanks, fixed it in two places it showed up in this file.  I will have to keep my eye out for this pattern when doing code cleanup.

mrry commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_maxpoolgrad.h:
```diff
@@ -334,38 +376,36 @@ class DnnlMaxPoolGrad : public DnnlKernel {
  private:
   dnnl::memory::format_tag GetAVXFormat(const dnnl::memory::dims& src_dims_mkl) {
     bool is_2D = src_dims_mkl.size() == 4 ? true : false;
     bool is_1D = src_dims_mkl.size() == 3 ? true : false;
```
Not in this PR, but I should have caught this last time... the ternary operator here is redundant.

```c++
bool is_2D = src_dims_mkl.size() == 4;
bool is_1D = src_dims_mkl.size() == 3;
```

georgen117 replied:
I can update this.

mrry reviewed (changes requested):
Thanks George... mostly just a few nits on my end and a few things I'd like to clarify.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:
```diff
@@ -67,12 +67,11 @@ class DnnlConvGrad : public DnnlKernel {
     dnnl::engine cpu_engine;
     dnnl::engine engine_to_use;
-    const auto iter = dnnl_engine.find(dnnl::engine::kind::cpu);
+    std::unordered_map<dnnl::engine::kind, dnnl::engine>::const_iterator iter = dnnl_engine.find(dnnl::engine::kind::cpu);
```
changing this to auto changes the iterator from a const_iterator to an iterator. The iterator is expected to be a const_iterator in the below code and using auto will cause a compilation error.

If OnnxRuntime suports C++17 then I could use `auto iter = std::as_const(dnnl_engine).find(dnnl::engine::kind::cpu);` I was not sure if OnnxRuntime supports C++17; so I went with the `const_iterator`.

mrry replied:
Thanks for explaining!

georgen117 commented on cmake/external/dnnl.cmake:
```diff
@@ -2,7 +2,7 @@ include (ExternalProject)
 
 set(DNNL_URL https://github.com/oneapi-src/onednn)
 # If DNNL_TAG is updated, check if MKLML_VERSION and platform.cmake.patch need to be updated.
```
Thanks for asking about this.  I double-checked the code and we are good with this change.

georgen117 commented:
@mrry is there anything I have missed? I feel like I have addressed all of your comments.  

There were a few of the nits pick changes that I chose not to make in this commit but want to make in a follow-up commit.

mrry commented:
Thanks George, this should be good to merge once all the CI pipelines pass.

georgen117 commented:
> Thanks George, this should be good to merge once all the CI pipelines pass.

Sounds great.  I have a cleanup commit that I was waiting on this commit before doing a pull commit.  There are a few of your nit-picks that will be incorporated in that upcoming change.

mrry commented:
/azp run Windows GPU CI Pipeline, Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline

mrry commented:
/azp run Windows GPU TensorRT CI Pipeline, centos7_cpu, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, Android CI Pipeline, Linux DNNL CI Pipeline, iOS CI Pipeline, orttraining-win-ci-pipeline, orttraining-win-gpu-ci-pipeline, orttraining-mac-ci-pipeline

mrry commented:
/azp run orttraining-distributed,orttraining-amd-gpu-ci-pipeline,Linux Nuphar CI Pipeline

georgen117 commented:
@mrry it looks like all of the checks have passed.  

Its looks like it's waiting on someone with write access to merge the pull request.

mrry commented:
Thanks for the reminder :). Merging now!

georgen117 commented:
> Thanks for the reminder :). Merging now!

Your welcome.  I figured you were just waiting for the CI builds to complete. Thanks again for the review work it really is appreciated.
